File: articles/calculus-for-systems-modeling/articles/power-series-and-functional-representation/fortran/power_series_audit.f90
program power_series_audit
  implicit none
  real(8), dimension(3) :: xs
  integer, dimension(3) :: ns
  real(8) :: x, partial, reference
  integer :: i, n, n_terms

  xs = (/0.25d0, 0.75d0, 1.25d0/)
  ns = (/5, 20, 10/)

  print '(A)', 'function_name center x_value n_terms partial_sum reference_value absolute_error convergence_status'

  do i=1,3
    x = xs(i)
    n_terms = ns(i)
    partial = 0.0d0
    do n=0,n_terms-1
      partial = partial + x**n
    end do

    if (abs(x) < 1.0d0) then
      reference = 1.0d0 / (1.0d0 - x)
      print '(A,1X,F5.1,1X,F8.4,1X,I4,1X,F14.8,1X,F14.8,1X,F14.8,1X,A)', '1/(1-x)', 0.0d0, x, n_terms, partial, reference, abs(reference-partial), 'inside_radius'
    else
      print '(A,1X,F5.1,1X,F8.4,1X,I4,1X,F14.8,1X,A)', '1/(1-x)', 0.0d0, x, n_terms, partial, 'outside_radius'
    end if
  end do
end program

File: articles/calculus-for-systems-modeling/articles/power-series-and-functional-representation/c/power_series_audit.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

double geometric_power_series(double x, int n_terms){
  double total = 0.0;
  for(int n=0; n<n_terms; n++){ total += pow(x, n); }
  return total;
}

int main(void){
  double xs[] = {0.25, 0.75, 1.25};
  int ns[] = {5, 20, 10};
  printf("function_name,center,x_value,n_terms,partial_sum,reference_value,absolute_error,convergence_status,warning\n");
  for(int i=0; i<3; i++){
    double x = xs[i];
    int n_terms = ns[i];
    double partial = geometric_power_series(x, n_terms);
    int converges = fabs(x) < 1.0;
    if(converges){
      double reference = 1.0 / (1.0 - x);
      printf("1/(1-x),0.0,%.6f,%d,%.12f,%.12f,%.12f,inside radius of convergence,\n",
             x, n_terms, partial, reference, fabs(reference - partial));
    } else {
      printf("1/(1-x),0.0,%.6f,%d,%.12f,,,outside radius of convergence,Power series does not converge for this x value.\n",
             x, n_terms, partial);
    }
  }
  return EXIT_SUCCESS;
}
